contrib/platform/src/com/sun/jna/platform/win32/VersionHelpers.java

```diff
@@ -0,0 +1,213 @@
+ * the project.
+ * 
+ * You may obtain a copy of the LGPL License at:
+ * 
+ * http://www.gnu.org/licenses/licenses.html
+ * 
+ * A copy is also included in the downloadable source code package
+ * containing JNA, in file "LGPL2.1".
+ * 
+ * You may obtain a copy of the Apache License at:
+ * 
+ * http://www.apache.org/licenses/
+ * 
+ * A copy is also included in the downloadable source code package
+ * containing JNA, in file "AL2.0".
+ */
+package com.sun.jna.platform.win32;
+
+import com.sun.jna.platform.win32.WinDef.DWORD;
+import com.sun.jna.platform.win32.WinDef.WORD;
+import com.sun.jna.platform.win32.WinNT.OSVERSIONINFOEX;
+
+/**
+ * The following functions can be used to determine the current operating system
+ * version or identify whether it is a Windows or Windows Server release. These
+ * functions provide simple tests that use the VerifyVersionInfo function and
+ * the recommended greater than or equal to comparisons that are proven as a
+ * robust means to determine the operating system version.
+ */
+public class VersionHelpers {
+
+    /**
+     * This function is useful in confirming a version of Windows Server that
+     * doesn't share a version number with a client release. You should only use
+     * this function if the other provided version helper functions do not fit
+     * your scenario.
+     * 
+     * @param wMajorVersion
+     *            The major version to test
+     * @param wMinorVersion
+     *            The minor version to test
+     * @param wServicePackMajor
+     *            The service pack to test
+     * @return True if the current OS version matches, or is greater than, the
+     *         provided version information.
+     */
+    public static boolean IsWindowsVersionOrGreater(int wMajorVersion, int wMinorVersion, int wServicePackMajor) {
+        OSVERSIONINFOEX osvi = new OSVERSIONINFOEX();
+        osvi.dwOSVersionInfoSize = new DWORD(osvi.size());
+        osvi.dwMajorVersion = new DWORD(wMajorVersion);
+        osvi.dwMinorVersion = new DWORD(wMinorVersion);
+        osvi.wServicePackMajor = new WORD(wServicePackMajor);
+
+        long dwlConditionMask = 0;
+        dwlConditionMask = Kernel32.INSTANCE.VerSetConditionMask(dwlConditionMask, WinNT.VER_MAJORVERSION,
+                WinNT.VER_GREATER_EQUAL);
+        dwlConditionMask = Kernel32.INSTANCE.VerSetConditionMask(dwlConditionMask, WinNT.VER_MINORVERSION,
+                WinNT.VER_GREATER_EQUAL);
+        dwlConditionMask = Kernel32.INSTANCE.VerSetConditionMask(dwlConditionMask, WinNT.VER_SERVICEPACKMAJOR,
+                WinNT.VER_GREATER_EQUAL);
+
+        return Kernel32.INSTANCE.VerifyVersionInfoW(osvi,
+                WinNT.VER_MAJORVERSION | WinNT.VER_MINORVERSION | WinNT.VER_SERVICEPACKMAJOR, dwlConditionMask);
+    }
+
+    /**
+     * @return true if the current OS version matches, or is greater than, the
+     *         Windows XP version.
+     */
+    public static boolean IsWindowsXPOrGreater() {
```

Comment:
I'm not sure whether I like this, but I don't see a saner version, that doesn't look like repeats...

What might be worth evaluation is, if pulling `HIBYTE(WinNT.WIN32_WINNT_WINXP)` and `LOBYTE(WinNT.WIN32_WINNT_WINXP)` into private static final variables might be worth it.

Reply:
I did try to faithfully reproduce the existing inline header file, in the interest of readability.  

I think adding variables would be worse than just explicitly doing the math (`>>8 & 0xFF`) and `(byte)` cast inline, which would be pre-calculated and optimized by the compiler.  (I expect the JIT optimizer would also eventually optimize the method calls as well if they were called frequently enough.)

Reply:
So you want me to go back to the HIBYTE and LOBYTE methods (that would be JIT optimized) rather than the current compiler-optimized bit math?  I kind of like the current version and think adding a comment would make it just as readable/understandable for future maintainers.

Reply:
My thinking was, that at this time I understand what you are doing there. But in 6-12 months I might have a WTF moment. I'm ok if you want to keep the bit shifts inline. 

Independend of this - my original idea was this:

```java
    private static final byte WIN32_WINNT_WINXP_MAJOR = HIBYTE(WinNT.WIN32_WINNT_WINXP);
    private static final byte WIN32_WINNT_WINXP_MINOR = LOBYTE(WinNT.WIN32_WINNT_WINXP);
    /**
     * @return true if the current OS version matches, or is greater than, the
     *         Windows XP version.
     */
    public static boolean IsWindowsXPOrGreater() {
        return IsWindowsVersionOrGreater(WIN32_WINNT_WINXP_MAJOR, WIN32_WINNT_WINXP_MINOR, 0);
    }
```

Pulling the calculation into a static final. But as said, if you like it this way, I won't stand in the way.

Reply:
I'll add a comment explaining what the bit shifts do, so you and I both can remember what's going on in 6-12 months.